plugins/uefi/generate-quirk.py

```diff
@@ -0,0 +1,70 @@
+#!/usr/bin/python3
+#
+# Copyright (C) 20202 Richard Hughes <richard@hughsie.com>
+#
+# SPDX-License-Identifier: LGPL-2.1+
+
+import os
+import sys
+import gzip
+import xml.etree.ElementTree as etree
+
+def get_kind_from_component(md):
+
+    # ideally we would use <category> but that's not available in the XML yet
+    name = md.find('name').text
+    if name.find('Embedded Controller') != -1:
+        return 'Embedded Controller'
+    if name.find('Corporate ME') != -1:
+        return 'Intel Management Engine'
+    if name.find('Consumer ME') != -1:
+        return 'Intel Management Engine'
+    return None
+
+def generate_quirk_from_xml(infn, outfn):
+
+    # parse GZipped XML
+    with gzip.open(infn, 'rb') as xml:
+        tree = etree.parse(xml)
+
+    # look for any interesting devices
+    seen_guids = {}
+    with open(outfn, 'w') as out:
+        out.write('# generated using {} -- DO NOT EDIT\n'.format(os.path.basename(sys.argv[0])))
+        for md in tree.findall('component'):
+
+            # filter by protocol
+            protocol = None
+            for val in md.findall('custom/value'):
+                if val.attrib['key'] == 'LVFS::UpdateProtocol':
+                    protocol = val.text
+            if protocol != 'org.uefi.capsule':
+                continue
+
+            # it is unlikely, but there might be multiple GUIDs here
+            guids = []
+            for prov in md.findall('provides/firmware'):
+                if prov.text in seen_guids:
+                    continue
+                guids.append(prov.text)
+            if not guids:
+                continue
+
+            # map to a suitable name for fwupd
+            kind = get_kind_from_component(md)
+            if not kind:
+                continue
+
+            # output another chunk
+            appstream_id = md.find('id').text
+            out.write('\n# from {}\n'.format(appstream_id))
+            for guid in guids:
+                out.write('[Guid={}]\n'.format(guid))
+                out.write('Name = {}\n'.format(kind))
```

Comment:
Do we perhaps want to also match `Vendor` here?  Although I know it's unlikely, it would help prevent a collision across vendors for ESRT GUIDs

Reply:
So you mean _match_, or _set_? I was hesitant to do the latter (and VendorId) and that seems a bit magic and undefined.

Reply:
Match, if it's possible.  I just want to avoid `95f9a8f4-9531-483c-afee-c3f12774ff4b` corresponds to an ME on Lenovo and SSD on Dell and wrong thing shows up.

Reply:
I don't think that's possible. If GUIDs did collide, I really would go and shout at some people pretty loud.

Reply:
Well the problem is no one gets to claim they own a GUID.  There is no public registry, nothing saying you need to clear your usage of a GUID or prefix with a standards body or anything.  So who do you yell at?  The second one who put firmware out?

On LVFS CHID are an optional lever to mitigate fallout from this happening.  But if it did happen then CHID doesn't apply to metadata like this.

I guess there is another way to go about this than what your PR is doing..
You can let the metadata coming _from LVFS_ **directly** influence the name of the device.  So before you refresh you see "UEFI Device firmware".  After you refresh you see "Intel Management Engine"

Reply:
> Well the problem is no one gets to claim they own a GUID

There are more GUIDs than stars in the universe. The likelihood of generating the same random GUID is just impossible. The only time I've heard it ever happening is things like databases that generate billions of GUIDs every minute. If we do clash GUIDs it's because some vendor has copied and pasted it. In that scenario the wrong device name is the last of their problems, and they'd have to offer a new system update to fix it before it would work.

> You can let the metadata coming from LVFS directly influence the name of the device

So far I've resisted doing this, as it's a bit of a slippery slope that starts with names, moves to flags, and before you know it there's a huge complex glob of code that's a huge weird state machine.

Reply:
> f we do clash GUIDs it's because some vendor has copied and pasted it.

Well copied and pasted it or **chose** to re-use it across two different platforms.  At least on WU that's "OK"; albeit really confusing.  I really wish the spec was better in this area.

> So far I've resisted doing this, as it's a bit of a slippery slope that starts with names, moves to flags, and before you know it there's a huge complex glob of code that's a huge weird state machine.

I do see how it's a slippery slope.  This PR as its stands today is not harmful.